File: src/lib/device.ts
/**
 * Device Protocol Integration - TypeScript Types and API
 * 
 * Provides type-safe access to device detection and protocol operations
 * for MTK BROM, Qualcomm EDL, Fastboot, Samsung Odin, and more.
 */

import { invoke } from '@tauri-apps/api/core';

// ─────────────────────────────────────────────────────────────────────────────
// Enums
// ─────────────────────────────────────────────────────────────────────────────

export enum DeviceMode {
  Brom = 'Brom',
  PreLoader = 'PreLoader',
  Edl = 'Edl',
  Fastboot = 'Fastboot',
  Adb = 'Adb',
  Mtp = 'Mtp',
  Recovery = 'Recovery',
  SamsungOdin = 'SamsungOdin',
  UnisocFdl = 'UnisocFdl',
  Unknown = 'Unknown'
}

export enum ProtocolType {
  MtkBrom = 'MtkBrom',
  MtkPreloader = 'MtkPreloader',
  QualcommEdl = 'QualcommEdl',
  Fastboot = 'Fastboot',
  SamsungOdin = 'SamsungOdin',
  UnisocFdl = 'UnisocFdl',
  Adb = 'Adb',
  Mtp = 'Mtp',
  Unknown = 'Unknown'
}

// ─────────────────────────────────────────────────────────────────────────────
// Interfaces
// ─────────────────────────────────────────────────────────────────────────────

export interface DetectedDevice {
  mode: DeviceMode;
  vid: number;
  pid: number;
  serial: string | null;
  manufacturer: string | null;
  product: string | null;
  bus: number;
  address: number;
  chipset: string | null;
  detectedAt: number;
}

export interface DeviceConnectionStatus {
  connected: boolean;
  device: DetectedDevice | null;
  protocol: ProtocolType;
  message: string;
}

export interface DeviceScanResult {
  devices: DeviceConnectionStatus[];
  count: number;
  hasSupported: boolean;
}

export interface FastbootDeviceInfo {
  serial: string;
  product: string;
  variant: string;
  bootloaderVersion: string;
  basebandVersion: string;
  secureBoot: boolean;
  unlocked: boolean;
}

export interface FlashProgress {
  written: number;
  total: number;
  percent: number;
  partition: string;
}

// ─────────────────────────────────────────────────────────────────────────────
// Device Detection API
// ─────────────────────────────────────────────────────────────────────────────

async function invokeCommand<T>(command: string, args?: Record<string, unknown>): Promise<T> {
  return await invoke<T>(command, args);
}

/**
 * Scan all USB devices and return detected devices
 */
export async function scanDevices(): Promise<DeviceScanResult> {
  return await invokeCommand<DeviceScanResult>('device_scan_all');
}

/**
 * Auto-detect and connect to the best available device
 */
export async function autoConnectDevice(): Promise<DeviceConnectionStatus> {
  return await invokeCommand<DeviceConnectionStatus>('device_auto_connect');
}

/**
 * Check if specific device mode is available
 */
export async function checkDeviceMode(mode: string): Promise<boolean> {
  return await invokeCommand<boolean>('device_check_mode', { mode });
}

/**
 * Get protocol type name for display
 */
export async function getProtocolName(protocol: string): Promise<string> {
  return await invokeCommand<string>('device_get_protocol_name', { protocol });
}

// ─────────────────────────────────────────────────────────────────────────────
// Fastboot API
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Detect if fastboot device is connected
 */
export async function fastbootDetect(): Promise<boolean> {
  return await invokeCommand<boolean>('fastboot_detect');
}

/**
 * Get fastboot device information
 */
export async function fastbootGetInfo(): Promise<FastbootDeviceInfo> {
  return await invokeCommand<FastbootDeviceInfo>('fastboot_get_info');
}

/**
 * Flash partition from file
 */
export async function fastbootFlashPartition(
  partition: string,
  filePath: string
): Promise<void> {
  return await invokeCommand<void>('fastboot_flash_partition', {
    partition,
    filePath
  });
}

/**
 * Erase partition
 */
export async function fastbootErasePartition(partition: string): Promise<void> {
  return await invokeCommand<void>('fastboot_erase_partition', { partition });
}

/**
 * Reboot device
 */
export async function fastbootReboot(): Promise<void> {
  return await invokeCommand<void>('fastboot_reboot');
}

/**
 * Reboot to bootloader
 */
export async function fastbootRebootBootloader(): Promise<void> {
  return await invokeCommand<void>('fastboot_reboot_bootloader');
}

/**
 * Reboot to recovery
 */
export async function fastbootRebootRecovery(): Promise<void> {
  return await invokeCommand<void>('fastboot_reboot_recovery');
}

/**
 * Unlock bootloader
 */
export async function fastbootUnlockBootloader(): Promise<void> {
  return await invokeCommand<void>('fastboot_unlock_bootloader');
}

/**
 * Lock bootloader
 */
export async function fastbootLockBootloader(): Promise<void> {
  return await invokeCommand<void>('fastboot_lock_bootloader');
}

// ─────────────────────────────────────────────────────────────────────────────
// Utility Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Get human-readable device mode name
 */
export function getDeviceModeName(mode: DeviceMode): string {
  const names: Record<DeviceMode, string> = {
    [DeviceMode.Brom]: 'MediaTek BROM',
    [DeviceMode.PreLoader]: 'MediaTek PreLoader',
    [DeviceMode.Edl]: 'Qualcomm EDL',
    [DeviceMode.Fastboot]: 'Fastboot',
    [DeviceMode.Adb]: 'Android ADB',
    [DeviceMode.Mtp]: 'MTP Device',
    [DeviceMode.Recovery]: 'Recovery Mode',
    [DeviceMode.SamsungOdin]: 'Samsung Odin',
    [DeviceMode.UnisocFdl]: 'UniSoc FDL',
    [DeviceMode.Unknown]: 'Unknown Device'
  };
  return names[mode];
}

/**
 * Get device mode color for UI
 */
export function getDeviceModeColor(mode: DeviceMode): string {
  const colors: Record<DeviceMode, string> = {
    [DeviceMode.Brom]: '#009688',
    [DeviceMode.PreLoader]: '#00BCD4',
    [DeviceMode.Edl]: '#9C27B0',
    [DeviceMode.Fastboot]: '#FF9800',
    [DeviceMode.Adb]: '#4CAF50',
    [DeviceMode.Mtp]: '#2196F3',
    [DeviceMode.Recovery]: '#FF5722',
    [DeviceMode.SamsungOdin]: '#1976D2',
    [DeviceMode.UnisocFdl]: '#795548',
    [DeviceMode.Unknown]: '#555555'
  };
  return colors[mode];
}

/**
 * Check if device is in download/flash mode
 */
export function isFlashMode(mode: DeviceMode): boolean {
  return [
    DeviceMode.Brom,
    DeviceMode.PreLoader,
    DeviceMode.Edl,
    DeviceMode.Fastboot,
    DeviceMode.SamsungOdin,
    DeviceMode.UnisocFdl
  ].includes(mode);
}

/**
 * Check if device is MTK-based
 */
export function isMtkDevice(mode: DeviceMode): boolean {
  return [DeviceMode.Brom, DeviceMode.PreLoader].includes(mode);
}

/**
 * Check if device is Qualcomm-based
 */
export function isQualcommDevice(mode: DeviceMode): boolean {
  return mode === DeviceMode.Edl;
}

File: src-tauri/src/commands/device.rs
use crate::device::{
    detector::{has_device_mode, DeviceMode},
    protocol_router::{scan_all_devices, DeviceConnectionStatus, ProtocolRouter, ProtocolType},
};
use serde::Serialize;

/// Unified device scan result
#[derive(Debug, Serialize)]
pub struct DeviceScanResult {
    pub devices: Vec<DeviceConnectionStatus>,
    pub count: usize,
    pub has_supported: bool,
}

/// Scan all USB devices and return detected devices
#[tauri::command]
pub async fn device_scan_all() -> Result<DeviceScanResult, String> {
    log::info!("[Device] Scanning all USB devices...");

    let devices = scan_all_devices()?;
    let count = devices.len();
    let has_supported = devices.iter().any(|d| d.connected);

    log::info!(
        "[Device] Found {} device(s), {} supported",
        count,
        has_supported
    );

    Ok(DeviceScanResult {
        devices,
        count,
        has_supported,
    })
}

/// Auto-detect and connect to the best available device
#[tauri::command]
pub async fn device_auto_connect() -> Result<DeviceConnectionStatus, String> {
    log::info!("[Device] Auto-detecting device...");

    let device = ProtocolRouter::auto_detect()?;
    let protocol = ProtocolRouter::mode_to_protocol(&device.mode);
    let message = ProtocolRouter::device_description(&device);
    let chipset = ProtocolRouter::identify_chipset(&device);

    log::info!("[Device] Connected: {} (chipset: {:?})", message, chipset);

    Ok(DeviceConnectionStatus {
        connected: true,
        device: Some(device),
        protocol,
        message,
    })
}

/// Check if specific device mode is available
#[tauri::command]
pub async fn device_check_mode(mode: String) -> Result<bool, String> {
    let target_mode = match mode.to_lowercase().as_str() {
        "brom" => DeviceMode::Brom,
        "preloader" => DeviceMode::PreLoader,
        "edl" => DeviceMode::Edl,
        "fastboot" => DeviceMode::Fastboot,
        "adb" => DeviceMode::Adb,
        "samsung" | "odin" => DeviceMode::SamsungOdin,
        "unisoc" | "fdl" => DeviceMode::UnisocFdl,
        "recovery" => DeviceMode::Recovery,
        _ => return Err(format!("Unknown device mode: {}", mode)),
    };

    has_device_mode(target_mode)
}

/// Get protocol type name for display
#[tauri::command]
pub async fn device_get_protocol_name(protocol: String) -> Result<String, String> {
    let proto_type = match protocol.to_lowercase().as_str() {
        "mtkbrom" => ProtocolType::MtkBrom,
        "mtkpreloader" => ProtocolType::MtkPreloader,
        "edl" | "qualcommedl" => ProtocolType::QualcommEdl,
        "fastboot" => ProtocolType::Fastboot,
        "samsungodin" => ProtocolType::SamsungOdin,
        "unisocfdl" => ProtocolType::UnisocFdl,
        "adb" => ProtocolType::Adb,
        "mtp" => ProtocolType::Mtp,
        "recovery" => ProtocolType::Recovery,
        _ => ProtocolType::Unknown,
    };

    Ok(format!("{:?}", proto_type))
}

/// Fastboot: Detect if fastboot device is connected
#[tauri::command]
pub async fn fastboot_detect() -> Result<bool, String> {
    crate::commands::fastboot::fastboot_device_connected()
}

/// Fastboot: Get device information
#[tauri::command]
pub async fn fastboot_get_info() -> Result<crate::commands::fastboot::FastbootDeviceInfo, String> {
    let session = crate::commands::fastboot::FastbootSession::open()?;
    session.get_variables()
}

/// Fastboot: Flash partition from file
#[tauri::command]
pub async fn fastboot_flash_partition(partition: String, file_path: String) -> Result<(), String> {
    let data = std::fs::read(&file_path)
        .map_err(|e| format!("Failed to read file '{}': {}", file_path, e))?;

    let session = crate::commands::fastboot::FastbootSession::open()?;
    session.flash_partition(&partition, &data)
}

/// Fastboot: Reboot device
#[tauri::command]
pub async fn fastboot_reboot() -> Result<(), String> {
    let session = crate::commands::fastboot::FastbootSession::open()?;
    session.reboot()
}

/// Fastboot: Reboot to bootloader
#[tauri::command]
pub async fn fastboot_reboot_bootloader() -> Result<(), String> {
    let session = crate::commands::fastboot::FastbootSession::open()?;
    session.reboot_bootloader()
}

/// Fastboot: Reboot to recovery
#[tauri::command]
pub async fn fastboot_reboot_recovery() -> Result<(), String> {
    let session = crate::commands::fastboot::FastbootSession::open()?;
    session.reboot_recovery()
}

/// Fastboot: Lock bootloader
#[tauri::command]
pub async fn fastboot_lock_bootloader() -> Result<(), String> {
    let session = crate::commands::fastboot::FastbootSession::open()?;
    session.lock_bootloader()
}
